be/src/exprs/agg/maxmin_by.h

```diff
@@ -0,0 +1,419 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+
+#pragma once
+
+#include <limits>
+#include <type_traits>
+
+#include "column/fixed_length_column.h"
+#include "column/type_traits.h"
+#include "exprs/agg/aggregate.h"
+#include "gutil/casts.h"
+#include "util/raw_container.h"
+namespace starrocks::vectorized {
+
+template <PrimitiveType PT, typename = guard::Guard>
+struct MaxByAggregateData {};
+
+template <PrimitiveType PT>
+struct MaxByAggregateData<PT, IntegralPTGuard<PT>> {
+    using T = RunTimeCppType<PT>;
+    raw::RawVector<uint8_t> buffer_result;
+    T max = std::numeric_limits<T>::lowest();
+    void reset() {
+        buffer_result.clear();
+        max = std::numeric_limits<T>::lowest();
+    }
+};
+
+template <PrimitiveType PT>
+struct MaxByAggregateData<PT, FloatPTGuard<PT>> {
+    using T = RunTimeCppType<PT>;
+    raw::RawVector<uint8_t> buffer_result;
+    T max = std::numeric_limits<T>::lowest();
+    void reset() {
+        buffer_result.clear();
+        max = std::numeric_limits<T>::lowest();
+    }
+};
+
+template <>
+struct MaxByAggregateData<TYPE_DECIMALV2, guard::Guard> {
+    raw::RawVector<uint8_t> buffer_result;
+    DecimalV2Value max = DecimalV2Value::get_min_decimal();
+    void reset() {
+        buffer_result.clear();
+        max = DecimalV2Value::get_min_decimal();
+    }
+};
+
+template <PrimitiveType PT>
+struct MaxByAggregateData<PT, DecimalPTGuard<PT>> {
+    using T = RunTimeCppType<PT>;
+    raw::RawVector<uint8_t> buffer_result;
+    T max = std::numeric_limits<T>::lowest();
+    void reset() {
+        buffer_result.clear();
+        max = std::numeric_limits<T>::lowest();
+    }
+};
+
+template <>
+struct MaxByAggregateData<TYPE_DATETIME, guard::Guard> {
+    raw::RawVector<uint8_t> buffer_result;
+    TimestampValue max = TimestampValue::MIN_TIMESTAMP_VALUE;
+    void reset() {
+        buffer_result.clear();
+        max = TimestampValue::MIN_TIMESTAMP_VALUE;
+    }
+};
+
+template <>
+struct MaxByAggregateData<TYPE_DATE, guard::Guard> {
+    raw::RawVector<uint8_t> buffer_result;
+    DateValue max = DateValue::MIN_DATE_VALUE;
+    void reset() {
+        buffer_result.clear();
+        max = DateValue::MIN_DATE_VALUE;
+    }
+};
+
+template <PrimitiveType PT, typename State, typename = guard::Guard>
+struct MaxByElement {
```

Comment:
You can reference other agg functions' implementation, for examples avg, distinct.
two suggestion:
1. MaxByAggregateData of  all types are the same, the difference is max value, it is depends on the concrete type, so max value should be abstracted into template, then MaxByAggregateData use this template directly.
2. the methods of MaxByElement should be the methods of MaxByAggregateData, MaxByElement can be eliminated.

Reply:
i referred to agg function min/max code, about MaxElement